src/Workspaces/Core/Desktop/Workspace/Host/TemporaryStorage/TemporaryStorageServiceFactory.ReferenceCountedDisposable.cs

```diff
@@ -0,0 +1,197 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Runtime.CompilerServices;
+using System.Threading;
+
+#if DEBUG
+using System.Diagnostics;
+#endif
+
+namespace Microsoft.CodeAnalysis.Host
+{
+    internal partial class TemporaryStorageServiceFactory
+    {
+        private sealed class ReferenceCountedDisposable<T> : IDisposable
+            where T : IDisposable
+        {
+            private T _instance;
+            private StrongBox<int> _referenceCount;
+
+            /// <summary>
+            /// Initializes a new reference counting wrapper around an <see cref="IDisposable"/> object.
+            /// </summary>
+            /// <remarks>
+            /// <para>The reference count is initialized to 1.</para>
+            /// </remarks>
+            /// <param name="instance">The object owned by this wrapper.</param>
+            public ReferenceCountedDisposable(T instance)
+            {
+                _instance = instance;
+                _referenceCount = new StrongBox<int>(1);
+            }
+
+            private ReferenceCountedDisposable(T instance, StrongBox<int> referenceCount)
+            {
+                _instance = instance;
+
+                // The reference count has already been incremented for this instance
+                _referenceCount = referenceCount;
+            }
+
+            /// <summary>
+            /// Gets the target object.
+            /// </summary>
+            /// <remarks>
+            /// <para>This call is not valid after <see cref="Dispose"/> is called.</para>
+            /// </remarks>
+            public T Target => _instance;
+
+            /// <summary>
+            /// Increments the reference count for the disposable object, and returns a new disposable reference to it.
+            /// </summary>
+            /// <remarks>
+            /// <para>The returned object is an independent reference to the same underlying object. Disposing of the
+            /// returned value multiple times will only cause the reference count to be decreased once.</para>
```

Comment:
> Why do we have a lock-free requirement?

Because it's not too challenging to get right at this scale.

> > which are now documented for the type.
>
> I'm not certain what this means.

Just means I added documentation in code explaining why it's implemented this way.